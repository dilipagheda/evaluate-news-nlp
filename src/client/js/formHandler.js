import { validateInput } from "./validateInput";
import { baseUrl } from "../constants/index";

function handleSubmit(event) {
  event.preventDefault();

  // check what text was put into the form field
  let formText = document.getElementById("name").value;
  if (!validateInput(formText)) {
    document.getElementById(
      "error-message"
    ).innerHTML = `Input text should not be empty or more than 5000 characters!`;
    return;
  }

  document.getElementById("error-message").innerHTML = "";
  document.getElementById("results").innerHTML = `<span>Please wait...</span>`;

  return fetch(`${baseUrl}/sentiment?text=${formText}`)
    .then((res) => res.json())
    .then(function (res) {
      updateUI(res, formText);
    })
    .catch((error) => {
      document.getElementById("results").innerHTML = error;
    });
}

function renderList(list) {
  const htmlListItems = [];

  if (list) {
    if (list.length > 0) {
      list.forEach((item) => {
        const htmlItem = `
                    <tr>
                        <td>${item.form}</td>
                        <td>${item.type}</td>
                        <td>${item.score_tag}</td>
                    </tr>
                `;
        htmlListItems.push(htmlItem);
      });
    }
  }

  if (htmlListItems.length > 0) {
    return `<table class='data-table'>
        <tr>
            <th>Form</th>
            <th>Type</th>
            <th>Score tag</th>
        </tr>
        ${htmlListItems.join(" ")}
    </table>`;
  } else {
    return "<span>No items found</span>";
  }
}

function updateUI(data, formText) {
  const sentimentedEntityList = `
    <div class='result-section'>
        <h2>Sentimented entity list</h2>
        ${renderList(data.sentimented_entity_list)}
    </div>
    `;

  const sentimentedConceptList = `
    <div class='result-section'>
        <h2>Sentimented concept list</h2>
        ${renderList(data.sentimented_concept_list)}
    </div>
    `;

  const htmlTable = `
    <div class="result-section">
        <h2>Overall results</h2>
        <table class='data-table'>
            <tr>
                <th>Score tag</th>
                <th>Agreement</th>
                <th>Confidence</th>
            </tr>
            <tr>
                <td>${data.score_tag}</td>
                <td>${data.agreement}</td>
                <td>${data.confidence}</td>
            </tr>
        </table>
    </div>
    `;

  const htmlOutput = `
        ${htmlTable}
        ${sentimentedEntityList}
        ${sentimentedConceptList}
    `;

  const htmlInput = `
        <h2>You entered</h2>
        <div class='user-input'>
            ${formText}
        </div>
        
    `;
  document.getElementById("user-input-section").innerHTML = htmlInput;
  document.getElementById("results").innerHTML = htmlOutput;
}

export { handleSubmit };
