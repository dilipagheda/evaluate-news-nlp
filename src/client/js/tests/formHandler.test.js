import { handleSubmit } from "../formHandler";
import { enableFetchMocks } from "jest-fetch-mock";

beforeEach(() => {
  enableFetchMocks();
  fetch.resetMocks();
});

const mockResponse = {
  status: {
    code: "0",
    msg: "OK",
    credits: "1",
    remaining_credits: "19981",
  },
  model: "general_en",
  score_tag: "P",
  agreement: "AGREEMENT",
  subjectivity: "OBJECTIVE",
  confidence: "100",
  irony: "NONIRONIC",
  sentence_list: [
    {
      text:
        'After the publication of our visual investigation on the Sydney "mega-blaze", readers asked us how we found the tree which started the Gospers Mountain bushfire',
      inip: "0",
      endp: "159",
      bop: "y",
      confidence: "100",
      score_tag: "P",
      agreement: "AGREEMENT",
      segment_list: [
        {
          text:
            'After the publication of our visual investigation on the Sydney "mega-blaze", readers asked us how we found the tree which started the Gospers Mountain bushfire',
          segment_type: "main",
          inip: "0",
          endp: "159",
          confidence: "100",
          score_tag: "P",
          agreement: "AGREEMENT",
          polarity_term_list: [
            {
              text: "start@V",
              inip: "123",
              endp: "129",
              confidence: "100",
              score_tag: "P",
              sentimented_concept_list: [
                {
                  form: "publication",
                  id: "403d3926ea",
                  variant: "publication",
                  inip: "10",
                  endp: "20",
                  type: "Top",
                  score_tag: "P",
                },
                {
                  form: "reader",
                  id: "030772557f",
                  variant: "readers",
                  inip: "78",
                  endp: "84",
                  type: "Top>Person",
                  score_tag: "P",
                },
                {
                  form: "tree",
                  id: "35f20d6542",
                  variant: "tree",
                  inip: "112",
                  endp: "115",
                  type: "Top>LivingThing>Flora",
                  score_tag: "P",
                },
              ],
            },
          ],
          sentimented_entity_list: [
            {
              form: "Sydney",
              id: "2a3fcc69a2",
              variant: "Sydney",
              inip: "57",
              endp: "62",
              type: "Top>Location>GeoPoliticalEntity>City",
              score_tag: "NONE",
            },
            {
              form: "Gospers Mountain",
              id: "__8908057099639966169",
              variant: "Gospers Mountain",
              inip: "135",
              endp: "150",
              type: "Top",
              score_tag: "NONE",
            },
          ],
          sentimented_concept_list: [
            {
              form: "investigation",
              id: "6e4db500b6",
              variant: "investigation",
              inip: "36",
              endp: "48",
              type: "Top",
              score_tag: "NONE",
            },
          ],
          segment_list: [
            {
              text: "which started the Gospers Mountain bushfire",
              segment_type: "main",
              inip: "117",
              endp: "159",
              confidence: "100",
              score_tag: "P",
              agreement: "AGREEMENT",
              polarity_term_list: [
                {
                  text: "start@V",
                  inip: "123",
                  endp: "129",
                  confidence: "100",
                  score_tag: "P",
                },
              ],
              sentimented_entity_list: [
                {
                  form: "Gospers Mountain",
                  id: "__8908057099639966169",
                  variant: "Gospers Mountain",
                  inip: "135",
                  endp: "150",
                  type: "Top",
                  score_tag: "NONE",
                },
              ],
            },
          ],
        },
      ],
      sentimented_entity_list: [
        {
          form: "Sydney",
          id: "2a3fcc69a2",
          type: "Top>Location>GeoPoliticalEntity>City",
          score_tag: "NONE",
        },
        {
          form: "Gospers Mountain",
          id: "__8908057099639966169",
          type: "Top",
          score_tag: "NONE",
        },
      ],
      sentimented_concept_list: [
        {
          form: "reader",
          id: "030772557f",
          type: "Top>Person",
          score_tag: "P",
        },
        {
          form: "tree",
          id: "35f20d6542",
          type: "Top>LivingThing>Flora",
          score_tag: "P",
        },
        {
          form: "publication",
          id: "403d3926ea",
          type: "Top",
          score_tag: "P",
        },
        {
          form: "investigation",
          id: "6e4db500b6",
          type: "Top",
          score_tag: "NONE",
        },
      ],
    },
    {
      text: "— a journalist explains how he did it.",
      inip: "161",
      endp: "198",
      bop: "n",
      confidence: "100",
      score_tag: "NONE",
      agreement: "AGREEMENT",
      segment_list: [
        {
          text: "— a journalist explains how he did it",
          segment_type: "secondary",
          inip: "161",
          endp: "197",
          confidence: "100",
          score_tag: "NONE",
          agreement: "AGREEMENT",
          polarity_term_list: [],
          sentimented_concept_list: [
            {
              form: "journalist",
              id: "0d2a86e544",
              variant: "journalist",
              inip: "165",
              endp: "174",
              type: "Top>OtherEntity>Vocation",
              score_tag: "NONE",
            },
          ],
        },
      ],
      sentimented_entity_list: [],
      sentimented_concept_list: [
        {
          form: "journalist",
          id: "0d2a86e544",
          type: "Top>OtherEntity>Vocation",
          score_tag: "NONE",
        },
      ],
    },
  ],
  sentimented_entity_list: [
    {
      form: "Sydney",
      id: "2a3fcc69a2",
      type: "Top>Location>GeoPoliticalEntity>City",
      score_tag: "NONE",
    },
    {
      form: "Gospers Mountain",
      id: "__8908057099639966169",
      type: "Top",
      score_tag: "NONE",
    },
  ],
  sentimented_concept_list: [
    {
      form: "reader",
      id: "030772557f",
      type: "Top>Person",
      score_tag: "P",
    },
    {
      form: "journalist",
      id: "0d2a86e544",
      type: "Top>OtherEntity>Vocation",
      score_tag: "NONE",
    },
    {
      form: "tree",
      id: "35f20d6542",
      type: "Top>LivingThing>Flora",
      score_tag: "P",
    },
    {
      form: "publication",
      id: "403d3926ea",
      type: "Top",
      score_tag: "P",
    },
    {
      form: "investigation",
      id: "6e4db500b6",
      type: "Top",
      score_tag: "NONE",
    },
  ],
};

test("verify handleSubmit function populates the sentiment results correctly", () => {
  const event = { preventDefault: () => {} };

  jest.spyOn(event, "preventDefault");

  fetch.mockResponse(() => {
    return new Promise((resolve) => {
      resolve(JSON.stringify(mockResponse));
    });
  });

  document.body.innerHTML = `
        <textarea id="name" name="input" rows="10"></textarea>
        <div id="error-message"></div>
        <section>
            <div id="user-input-section"></div>
        </section>

        <section>
            <div id="results"></div>
        </section>       
    `;

  document.getElementById("name").value = "hi there";

  return handleSubmit(event).then(() => {
    //verify that preventDefault is called
    expect(event.preventDefault).toBeCalled();

    //Verify that portion of response data is in DOM
    expect(document.body).toMatchSnapshot();
  });
});

test("verify handleSubmit function displays the message please wait.. while waiting for the result from api", () => {
  const event = { preventDefault: () => {} };

  jest.spyOn(event, "preventDefault");

  fetch.mockResponse(() => {
    //Verify that portion of response data is in DOM
    expect(document.body).toMatchSnapshot();
    return new Promise((resolve) => {
      resolve(JSON.stringify(mockResponse));
    });
  });

  document.body.innerHTML = `
        <textarea id="name" name="input" rows="10"></textarea>
        <div id="error-message"></div>
        <section>
            <div id="user-input-section"></div>
        </section>

        <section>
            <div id="results"></div>
        </section>       
    `;

  document.getElementById("name").value = "hi there";

  return handleSubmit(event).then(() => {
    //verify that preventDefault is called
    expect(event.preventDefault).toBeCalled();
  });
});

test("verify handleSubmit function displays the error for invalid input text and fetch call is not made", () => {
  const event = { preventDefault: () => {} };

  jest.spyOn(event, "preventDefault");
  jest.spyOn(global, "fetch");

  document.body.innerHTML = `
      <textarea id="name" name="input" rows="10"></textarea>
      <div id="error-message"></div>
      <section>
          <div id="user-input-section"></div>
      </section>

      <section>
          <div id="results"></div>
      </section>       
  `;

  document.getElementById("name").value = ""; //pass empty value

  handleSubmit(event);
  expect(event.preventDefault).toBeCalled();
  expect(global.fetch).not.toBeCalled();

  expect(document.body).toMatchSnapshot();
});
