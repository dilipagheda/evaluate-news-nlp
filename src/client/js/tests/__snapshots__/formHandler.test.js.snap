// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify handleSubmit function displays the error for invalid input text and fetch call is not made 1`] = `
<body>
  
      
  <textarea
    id="name"
    name="input"
    rows="10"
  />
  
      
  <div
    id="error-message"
  >
    Input text should not be empty or more than 5000 characters!
  </div>
  
      
  <section>
    
          
    <div
      id="user-input-section"
    />
    
      
  </section>
  

      
  <section>
    
          
    <div
      id="results"
    />
    
      
  </section>
         
  
</body>
`;

exports[`verify handleSubmit function displays the message please wait.. while waiting for the result from api 1`] = `
<body>
  
        
  <textarea
    id="name"
    name="input"
    rows="10"
  />
  
        
  <div
    id="error-message"
  />
  
        
  <section>
    
            
    <div
      id="user-input-section"
    />
    
        
  </section>
  

        
  <section>
    
            
    <div
      id="results"
    >
      <span>
        Please wait...
      </span>
    </div>
    
        
  </section>
         
    
</body>
`;

exports[`verify handleSubmit function populates the sentiment results correctly 1`] = `
<body>
  
        
  <textarea
    id="name"
    name="input"
    rows="10"
  />
  
        
  <div
    id="error-message"
  />
  
        
  <section>
    
            
    <div
      id="user-input-section"
    >
      
        
      <h2>
        You entered
      </h2>
      
        
      <div
        class="user-input"
      >
        
            hi there
        
      </div>
      
        
    
    </div>
    
        
  </section>
  

        
  <section>
    
            
    <div
      id="results"
    >
      
        
    
      <div
        class="result-section"
      >
        
        
        <h2>
          Overall results
        </h2>
        
        
        <table
          class="data-table"
        >
          
            
          <tbody>
            <tr>
              
                
              <th>
                Score tag
              </th>
              
                
              <th>
                Agreement
              </th>
              
                
              <th>
                Confidence
              </th>
              
            
            </tr>
            
            
            <tr>
              
                
              <td>
                P
              </td>
              
                
              <td>
                AGREEMENT
              </td>
              
                
              <td>
                100
              </td>
              
            
            </tr>
            
        
          </tbody>
        </table>
        
    
      </div>
      
    
        
    
      <div
        class="result-section"
      >
        
        
        <h2>
          Sentimented entity list
        </h2>
        
        
        <table
          class="data-table"
        >
          
        
          <tbody>
            <tr>
              
            
              <th>
                Form
              </th>
              
            
              <th>
                Type
              </th>
              
            
              <th>
                Score tag
              </th>
              
        
            </tr>
            
        
                    
            <tr>
              
                        
              <td>
                Sydney
              </td>
              
                        
              <td>
                Top&gt;Location&gt;GeoPoliticalEntity&gt;City
              </td>
              
                        
              <td>
                NONE
              </td>
              
                    
            </tr>
            
                 
                    
            <tr>
              
                        
              <td>
                Gospers Mountain
              </td>
              
                        
              <td>
                Top
              </td>
              
                        
              <td>
                NONE
              </td>
              
                    
            </tr>
            
                
    
          </tbody>
        </table>
        
    
      </div>
      
    
        
    
      <div
        class="result-section"
      >
        
        
        <h2>
          Sentimented concept list
        </h2>
        
        
        <table
          class="data-table"
        >
          
        
          <tbody>
            <tr>
              
            
              <th>
                Form
              </th>
              
            
              <th>
                Type
              </th>
              
            
              <th>
                Score tag
              </th>
              
        
            </tr>
            
        
                    
            <tr>
              
                        
              <td>
                reader
              </td>
              
                        
              <td>
                Top&gt;Person
              </td>
              
                        
              <td>
                P
              </td>
              
                    
            </tr>
            
                 
                    
            <tr>
              
                        
              <td>
                journalist
              </td>
              
                        
              <td>
                Top&gt;OtherEntity&gt;Vocation
              </td>
              
                        
              <td>
                NONE
              </td>
              
                    
            </tr>
            
                 
                    
            <tr>
              
                        
              <td>
                tree
              </td>
              
                        
              <td>
                Top&gt;LivingThing&gt;Flora
              </td>
              
                        
              <td>
                P
              </td>
              
                    
            </tr>
            
                 
                    
            <tr>
              
                        
              <td>
                publication
              </td>
              
                        
              <td>
                Top
              </td>
              
                        
              <td>
                P
              </td>
              
                    
            </tr>
            
                 
                    
            <tr>
              
                        
              <td>
                investigation
              </td>
              
                        
              <td>
                Top
              </td>
              
                        
              <td>
                NONE
              </td>
              
                    
            </tr>
            
                
    
          </tbody>
        </table>
        
    
      </div>
      
    
    
    </div>
    
        
  </section>
         
    
</body>
`;
